### PR TITLE
ARROW-12556: [C++][Gandiva] Implement BYTESUBSTRING function on Gandiva

### PR DESCRIPTION
**Pull request by jpedroantunes**
Implement BYTE_SUBSTR([string] giventext, [number] x, [number] y)

Produces the binary representation of a string y characters long derived by starting at position x in the string giventext. y may also be given by the expression LENGTH(giventext), which indicates that you wish to convert every remaining character in giventext.

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -1520,4 +1520,40 @@ const char* binary_string(gdv_int64 context, const char* text, gdv_int32 text_le
...
+      reinterpret_cast<gdv_binary>(gdv_fn_context_arena_malloc(context, text_len));
+
+  if (ret == nullptr) {
+    gdv_fn_context_set_error_msg(context, "Could not allocate memory for output string");
+    *out_len = 0;
+    return "";
+  }
+
+  int32_t startPos;
+  if (offset < 0) {
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops.cc:1543]: Looks like it offset can be negative in which case it starts from the end. Can you add these cases in a comment at the top of the function

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 2]
```diff
@@ -1520,4 +1520,40 @@ const char* binary_string(gdv_int64 context, const char* text, gdv_int32 text_le
...
+    startPos = text_len + offset;
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops.cc:1544]: You need to validate the offset or else startPos may go negative

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 3]
```diff
@@ -1520,4 +1520,40 @@ const char* binary_string(gdv_int64 context, const char* text, gdv_int32 text_le
...
+  } else {
+    startPos = offset - 1;
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops.cc:1546]: Write a comment at the top of the function that offset is 1-indexed

File: cpp/src/gandiva/function_registry_string.cc [hunk 4]
```diff
@@ -216,6 +216,10 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
                      "concat_utf8_utf8_utf8_utf8_utf8_utf8_utf8_utf8_utf8_utf8",
                      NativeFunction::kNeedsContext),
 
+      NativeFunction("byte_substr", {"bytesubstring"},
+                     DataTypeVector{binary(), int32(), int32()}, binary(),
+                     kResultNullIfNull, "byte_substr", NativeFunction::kNeedsContext),
```
Comment (projjal) [inline cpp/src/gandiva/function_registry_string.cc:221]: nit: other function signature keep the argument types in the function name. Lets keep that convention

**Review by projjal — APPROVED** [review]

**Comment by praveenbingo** [comment]
@projjal needs a merge conflict

**Comment by praveenbingo** [comment]
@projjal can you please approve the latest..

**Review by projjal — APPROVED** [review]

**Comment by projjal** [comment]
> @projjal can you please approve the latest..

@praveenbingo done

**Pull request closed without merging**